enhancements/windows-containers/bring-your-own-windows-host.md

```diff
@@ -0,0 +1,171 @@
+* Handle upgrading of the Windows operating system
+* OpenShift Builds
+* OpenShift DeploymentConfigs
+
+## Proposal
+
+The cluster admin will create a ConfigMap called _windows-instances_ in the
+_openshift-windows-machine-config-operator_ namespace. WMCO will be configured
+to watch for this special ConfigMap, whose data section will have the following
+layout:
+```yaml
+kind: ConfigMap
+apiVersion: v1
+metadata:
+  name: windows-instances
+  namespace: openshift-windows-machine-config-operator
+data:
+  instance.1: |-
+    username=Administrator
+    ip-address=10.1.42.1
+  instance.2: |-
+    username=core
+    ip-address=10.1.42.2
+```
+WMCO will configure or de-configure instances based on the contents of the data
+section and will use the `cloud-private-key` to communicate with them.
+
+### User Stories
+
+#### Story 1
+As an OpenShift cluster administrator I would like to add an existing Windows
+instance to the cluster as a worker node.
+
+#### Story 2
+As an OpenShift cluster administrator I would like to remove an existing Windows
+instance from the cluster as a worker node.
+
+#### Story 3
+As an OpenShift cluster administrator I would like to have the Kubernetes
+components on Windows nodes upgraded when new Kubernetes versions or patch
+versions are released.
+
+### Risks and Mitigations
+
+The same risk and mitigation that apply to Windows nodes added using MachineSets
+to an IPI cluster applies here.
+
+## Design Details
+
+Changes to the _windows-instances_ ConfigMap will cause WMCO to compare the
+list of instances with the set of existing Windows nodes. The comparison will
+be based on the instance and node's IP addresses. WMCO will then react in the
+following manner:
+* If a new entry is found i.e. the list of nodes does not contain one that map
+  to the instances' IP address, the instance will be configured by WMCO. The
+  configuration process will be identical to configuring instances added using
+  a MachineSet in an IPI cluster.
+* If the list of nodes has IP addresses that do not map to any in the list of
+  instances, WMCO will perform an uninstall operation on the node. This will
+  involve:
+  * Drain the node
+  * Stop and remove
+    * Monitoring and logging services
+    * kube-proxy service
+    * hybrid-overlay-node service
+    * kubelet service
+  * Delete the payload binaries
+* If the list of nodes has IP addresses that do not map to any in the list of
+  instances, and WMCO is unable to access the instance, it will assume that
+  the node in question is not managed by it.
+
+The status of these operations will be reflected by events in the
+_openshift-windows-machine-config-operator_ namespace and will be associated
+with the _windows-instances_ ConfigMap.
+
+The upgrade process for BYOH nodes will be identical to upgrading Windows nodes
+added using a MachineSet in an IPI cluster.
+
+### Test Plan
+
+The existing set of WMCO end to end tests will be run against already created
```

Comment:
We also have machine deletion logic in test suite which can bring in machine-api into play if it is IPI which does the automatic drain of nodes or spawns up new machine when the node is not in running state for a long time

Reply:
Good point. We might need to use cloud APIs to bring up the VMs (oh no WNI again :-))

Alternatively stop the machine-api controller using CVO overrides after the VM is up. We would do that only when running the e2e for BYOH. Bit of a hack but I might be willing to live with it given it is an e2e test.